c10/cuda/CUDAException.cpp

```diff
@@ -1,13 +1,96 @@
 #include <c10/cuda/CUDAException.h>
 
+#include <c10/util/Backtrace.h>
 #include <c10/util/Exception.h>
+#include <c10/util/irange.h>
 #include <cuda_runtime.h>
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
 #include <string>
 
 namespace c10 {
 namespace cuda {
 
+/// Check that kernels ran correctly by checking the message buffer. BLOCKING.
+std::string c10_retrieve_device_side_assertion_info() {
+  std::stringstream oss;
+
+#ifdef TORCH_USE_CUDA_DSA
+  // The snapshot causes a brief block
+  const auto launch_data =
+      CUDAKernelLaunchRegistry::get_singleton_ref().snapshot();
+  const auto& assertion_data = launch_data.first;
+  const auto& launch_infos = launch_data.second;
+
+  oss << "# of GPUs this process interacted with = "
+      << CUDAKernelLaunchRegistry::get_singleton_ref().gpus_interacted_with()
+      << std::endl;
+
+  // Loop over each device that could be managed by the process
+  for (const auto device_num : c10::irange(assertion_data.size())) {
+    const auto& assertion_data_for_device = assertion_data.at(device_num);
+
+    // Did anything fail?
+    const auto failures_found = std::min(
+        assertion_data_for_device.assertion_count,
+        C10_CUDA_DEVICE_SIDE_ASSERTION_COUNT);
+    if (failures_found == 0) {
+      continue;
+    }
+
+    // Something failed, let's talk about that
+    oss << failures_found
+        << " CUDA device-side assertion failures were found on GPU #"
+        << device_num << "!" << std::endl;
+    oss << "But at least " << assertion_data_for_device.assertion_count
```

Comment:
should this be included only if number of asserts is more than C10_CUDA_DEVICE_SIDE_ASSERTION_COUNT?

Reply:
I've incorporated your suggestion, thanks.